enhancements/cloud-integration/openstack/failure-domains.md

```diff
@@ -0,0 +1,741 @@
+
+Note that the above is subject to change due to upstream review.
+
+### Workflow Description
+
+#### Create a new cluster with control plane and worker failure domains
+
+An OpenStack cloud has 3 failure domains: az1, az2, and az3. Each of these has corresponding Nova(compute) and Cinder(storage) availability zones of the same name. Additionally each failure domain contains a pre-created provider network with a subnet.
+
+The user wants to create a control plane with 3 nodes with each node in a different failure domain, and 6 workers distributed evenly across the same 3 failure domains.
+
+The user creates an install config containing the following:
+
+```yaml
+platform:
+  openstack:
+    machinesSubnet: 31ad4088-7e52-48da-9df5-35e70da438f6
+controlPlane:
+  name: control-plane
+  platform:
+    openstack:
+      type: m1.large
+      rootVolume:
+        size: 50
+        type: gold
+      failureDomains:
+      - computeAvailabilityZone: az1
+        storageAvailabilityZone: az1
+        ports:
+        - networkID: 4a270cf8-f5f9-4a09-8cfc-d39d77aa2ba1
+          fixedIPs:
+          - subnetID: 31ad4088-7e52-48da-9df5-35e70da438f6
+          tags:
+          - "control-plane"
+      - computeAvailabilityZone: az2
+        storageAvailabilityZone: az2
+        ports:
+        - networkID: dbc8664c-feee-4168-854b-75f8eb8eb258
+          fixedIPs:
+          - subnetID: c23f88c0-910a-4d4a-9ca6-9a11a0d71f40
+          tags:
+          - "control-plane"
+      - computeAvailabilityZone: az3
+        storageAvailabilityZone: az3
+        ports:
+        - networkID: a996c932-2aa1-4095-a40e-765bf844fcb4
+          fixedIPs:
+          - subnetID: 61d376b0-3714-4549-b017-3a5e0007f07c
+          tags:
+          - "control-plane"
+  replicas: 3
+compute:
+- name: worker
+  platform:
+    openstack:
+      type: m1.large
+      rootVolume:
+        size: 50
+        type: silver
+      failureDomains:
+      - computeAvailabilityZone: az1
+        storageAvailabilityZone: az1
+        ports:
+        - networkID: 4a270cf8-f5f9-4a09-8cfc-d39d77aa2ba1
+          fixedIPs:
+          - subnetID: 31ad4088-7e52-48da-9df5-35e70da438f6
+          tags:
+          - "control-plane"
+      - computeAvailabilityZone: az2
+        storageAvailabilityZone: az2
+        ports:
+        - networkID: dbc8664c-feee-4168-854b-75f8eb8eb258
+          fixedIPs:
+          - subnetID: c23f88c0-910a-4d4a-9ca6-9a11a0d71f40
+          tags:
+          - "control-plane"
+      - computeAvailabilityZone: az3
+        storageAvailabilityZone: az3
+        ports:
+        - networkID: a996c932-2aa1-4095-a40e-765bf844fcb4
+          fixedIPs:
+          - subnetID: 61d376b0-3714-4549-b017-3a5e0007f07c
+          tags:
+          - "control-plane"
+  replicas: 6
+```
```

Comment:
one thing that is not covered (yet) in this enhancement, could you please clarify whether we'll keep `machinesSubnet`? I think there is an opportunity to get rid of it, and maybe unify BYON use-case with how we'll do Failure Domains.

Of course `machinesSubnet` can't be removed for backward compatibility but underneath we could create a default failure domain that will use ports from this subnet.

Reply:
See https://github.com/openshift/installer/pull/6797 where we try to add `machinesSubnets`, whereas with FailureDomains, we could do something like:

```yaml
failureDomain:
  ports:
  - networkID: <Dual stack network ID>
    fixedIPs:
    - subnetID: <IPv4 subnet ID>
    - subnetID: <IPv6 subnet ID>
    tags:
    - "control-plane"
```

Reply:
To me, both `machinesSubnet` & FailureDomains help with BYON, but FailureDomains brings deeps evolution as it's described here, so I think it's worth discussing it.

Reply:
I tried to explain here: https://github.com/openshift/enhancements/pull/1319/files#diff-f27af7058fddb21715977065ad74f5ad53b668cbcf00c8cc971b12b1e9bf991fR219-R221

Can you help me make it clearer?

Also we will probably have to mention `machinesSubnets` with an S if dual-stack merges before us.

Reply:
works for me as well. We continue with the machinesSubnet field